# Working at the exegetical workbench
## Setting up your workbench
The most efficient way of working with Biblical Hebrew texts is by using *three* **tools**:
    
| Tool    | Function   |
| :---    | :---       |
| **The Library**: commerical `BibleSoftware` (Logos, Accordance)                                                          | great for looking up dictionaries, commentaries, and annotating your text (notes, highlighting, etc.)                 |
| **The Text** with linguistic annotations on all language levels (phonology, morphology, syntax, text-grammar): `SHEBANQ` | great for understanding the linguistic structure of a text and registering database codes useful for building queries |
| **The Analytic Tool-Kit**: `Text-Fabric` (i.e. SHEBANQ-query-tool on steroids)                                           | great for turning your exegetical questions into queries AND for analyzing your query results                       |




## `The Texts` in SHEBANQ
Lets pull up Gen 20 in SHEBANQ. Do the following:
1. Go to https://shebanq.ancient-data.org/ 
2. Click on "Text" in the upper left corner
3. Maneuver to Gen 20
4. Visualize the text-grammatical structure of Gen 20 by clicking on the "Notes (1/3)" button until it shows "Syntax (2/3)". You should see something like this:

![SHEBANQ Text-Grammar](https://eer8lq.dm.files.1drv.com/y4mqhs-4G3y3LkO0QP-D8de24hlFXrJuGEAxUd5GR5kTzGh_-iA8Yn1tLflAQHteLH1cTxa1a5foO5AUkcImzXUQ5PKMQF-FYtou40iGIWxgeJtqiS6QoFNRpMAR68Jju9X4pKH9eBUfOXV13uTfPePKJ133Lxh2ZzM8EwMd4gH0P1s6gp0kV8nUqXwsAvFacD1xXCyip3NnSFZ1dWTBinSnQ/Annotation%202019-06-11%20072101.png)

5. Study the codes and make sense out of them.

### Searching WayX clauses
Obviously `WayX` clauses are those clauses that are built by Wayyiqtol verbal forms followed by an explicit subject (marked by "X"). Lets search for those WayX clauses in Gen 20.

Before we can run our TF queries lets first run the preparatory codes:

In [4]:
%load_ext autoreload
%autoreload 2

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


In [5]:
# First, I have to laod different modules that I use for analyzing the data and for plotting:
import sys, os, collections
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt; plt.rcdefaults()
from matplotlib.pyplot import figure
from collections import Counter

# Second, I have to load the Text Fabric app
from tf.fabric import Fabric
from tf.app import use

In [6]:
A = use('bhsa', hoist=globals())

	connecting to online GitHub repo annotation/app-bhsa ... connected


GithubException: 504 {'data': "<html><body><h1>504 Gateway Time-out</h1>\nThe server didn't respond in time.\n</body></html>\n"}

Now, lets search for WayX clauses in Gen 20

In [16]:
WayXinGen20 = '''
verse book=Genesis chapter=20
    clause typ=WayX
'''
WayXinGen20  = A.search(WayXinGen20)
A.table(WayXinGen20, start=1, end=40, condensed=True, colorMap={1: 'blue', 2: 'blue', 3: 'blue', 4: 'blue'})

  0.12s 15 results


### Searching for proper names in Gen 20
Who are the participants in our text? Lets look for `nmpr` (nominal proper names). A complete list of part of speech (`sp`) can be found here: https://etcbc.github.io/bhsa/features/sp/

code|description
---|---
`art`  |article
`verb` |verb
`subs` |noun
`nmpr` |proper noun
`advb` |adverb
`prep` |preposition
`conj` |conjunction
`prps` |personal pronoun
`prde` |demonstrative pronoun
`prin` |interrogative pronoun
`intj` |interjection
`nega` |negative particle
`inrg` |interrogative particle
`adjv` |adjective

In [17]:
ProperNames = '''
verse book=Genesis chapter=20
    word sp=nmpr
'''
ProperNames  = A.search(ProperNames)
A.table(ProperNames, start=1, end=40, condensed=True, colorMap={1: 'yellow', 2: 'yellow', 3: 'yellow', 4: 'yellow'})

  0.48s 29 results


In [18]:
# Task 1
# search for the explicit subjects in those WayX clauses
# in order to find the right code for phrase functions go here: https://etcbc.github.io/bhsa/features/function/

We can see that while the story is all about Sara she does only seem to appear as a syntactical object and never as a syntactical subject in this narrative. She seems to be "objectified". Lets double check by doing a count of all lexemes that appear in subject position. For simplicity lets define that we allow only one word to appear in the subject phrase by using the operator `::`.

In [19]:
Subjects = '''
verse book=Genesis chapter=20
    clause typ=WayX
        phrase function=Subj
            :: word lex_utf8*
'''
Subjects  = A.search(Subjects)
A.table(Subjects, start=1, end=2, condensed=True, colorMap={1: 'yellow', 2: 'yellow', 3: 'yellow', 4: 'yellow'})

  1.33s 11 results


In order to use `pandas` count function we need to export our results as tsv file and then import it as a pandas dataframe. The cells below show my own folder in which I save my results. Since your folder system is different you need to alter the code so that it works with your own chosen folder.

In [22]:
A.export(Subjects, toDir='/Users/JOTA/Desktop/', toFile='Subjects.tsv')

In [23]:
Subjects=pd.read_csv('/Users/JOTA/Desktop/Subjects.tsv',delimiter='\t',encoding='utf-16')
pd.set_option('display.max_columns', 50)
Subjects.head(2)

R       S1  S2  S3    NODE1  TYPE1  \
0  1  Genesis  20   1  1414850  verse   
1  2  Genesis  20   2  1414851  verse   

                                               TEXT1    book1  chapter1  \
0  וַיִּסַּ֨ע מִשָּׁ֤ם אַבְרָהָם֙ אַ֣רְצָה הַנֶּ֔...  Genesis        20   
1  וַיֹּ֧אמֶר אַבְרָהָ֛ם אֶל־שָׂרָ֥ה אִשְׁתֹּ֖ו א...  Genesis        20   

    NODE2   TYPE2                                              TEXT2  typ2  \
0  429280  clause  וַיִּסַּ֨ע מִשָּׁ֤ם אַבְרָהָם֙ אַ֣רְצָה הַנֶּ֔...  WayX   
1  429283  clause      וַיֹּ֧אמֶר אַבְרָהָ֛ם אֶל־שָׂרָ֥ה אִשְׁתֹּ֖ו   WayX   

    NODE3   TYPE3        TEXT3 function3  NODE4 TYPE4        TEXT4 lex_utf84  
0  656856  phrase  אַבְרָהָם֙       Subj   9250  word  אַבְרָהָם֙      אברהם  
1  656866  phrase  אַבְרָהָ֛ם       Subj   9267  word  אַבְרָהָ֛ם      אברהם

In [24]:
Subjects["lex_utf84"].value_counts()

אבימלך    5
אברהם     4
אלהים     2
Name: lex_utf84, dtype: int64

Our assumption is confirmed. Sara is absent as an actor! Lets check whether she is present as an object (`Objc`) and indirect object (`Cmpl`):

In [25]:
# Task 2
# Go to SHEBANQ and look for how the lex of Sara looks like. 
# Then search for all cases in which Sara appears in Gen 20.
# In what syntactical position does she appear when she is explicitly mentioned?

## ETCBC database structure
### general overview
After we have executed some queries we have become aware that there are many codes for different linguistic levels. To really know what we are doing and searching for, we need to get a better understanding of the ETCBC database.

| **object types**       | **feature** | **value**              |
|:---------------------- |:------------|:-----------------------|
|text level: verse       | book        | Genesis, etc.          |
|   "                    | chapter     | 1, 2, etc.             |
|   "                    | verse       | 1, 2, etc.             |
|   "                    | etc.        | 1, 2, etc.             |
|syntax level: sentence  | sentence    |                        |
|syntax level: clause    | typ         | WayX, Way0, etc.       |
|   "                    | rela        | Attr, Objc, etc.       |
|   "                    | domain      | N, Q, etc.             |
|   "                    | etc.        |                        |
|syntax level: phrase    | typ         | VP, NP, etc.           |
|   "                    | function    | Pred, Subj, Objc, etc. |
|   "                    | etc.        |                        |
|word level: word        | lex         | "\>BRHM/", etc.        |
|   "                    | lex_utf8    | "אברהם/", etc.          |
|   "                    | gn          | m, f                   |
|   "                    | nu          | sg, pl, du             |
|   "                    | ps          | p1, p2, p3             |
|   "                    | st          | a, c                   |
|   "                    | vs          | qal, nif, etc.         |
|   "                    | vt          | perf, impf, etc.       |
|   "                    | etc.        |                        |

Thus, you can search on any of the above mentioned levels (word, phrase, clause, sentence, verse) and do any combination of those levels. Lets see some examples:

In [26]:
wordlevel='''
word lex=>BRHM/
'''
wordlevel  = A.search(wordlevel)
A.table(wordlevel, start=1, end=2, condensed=True)

  0.66s 175 results


In [27]:
phraselevel='''
phrase function=Subj
'''
phraselevel  = A.search(phraselevel)
A.table(phraselevel, start=1, end=2, condensed=True)

  0.72s 31952 results


In [28]:
clauselevel='''
clause rela=Attr
'''
clauselevel  = A.search(clauselevel)
A.show(clauselevel, start=1, end=2, condensed=True)

  0.23s 5809 results


In [29]:
textlevel='''
verse book=Genesis chapter=1 verse=1
'''
textlevel  = A.search(textlevel)
A.show(textlevel, start=1, end=2, condensed=True)

  0.06s 1 result


In [30]:
combinedlevels='''
verse book=Genesis
    clause rela=Attr
        phrase function=Subj
            word lex=>LHJM/
'''
combinedlevels  = A.search(combinedlevels)
A.show(combinedlevels, start=1, end=2, condensed=True)

  1.18s 12 results


### Closer Look into the database
Lets have a look at the Gen 20:1-4 on a text-grammatical level:

**The so called `CTT` file with Hebrew fonts with `rev`post-production:**
![](https://cp0uaa.dm.files.1drv.com/y4mmbHDW5LtgtcpKqVlWK1veylaLj2SF1C3PkjLyh6C_7Mdq45w42vcv2dSzID9bTQWlsMx2FnfSfQUufG6w09i9HLg2rN714eC86lSAkJ1VshtEQ-92c5s-1od6a09Xy9xfkTRhSRBpYnq0xrjZdmh5z3tc3zPDN6BK_MUxd7z-PRdfcHLjX_FbAOkA5qmdl495-ZIi0KRI_s7qKYkqwD1og/Annotation%202019-06-11%20114813.png)


**And here the original database `CTT` file without post-production:**
![](https://qlsuua.dm.files.1drv.com/y4mf7iN6v2x4XyfzUv8YXrEnVOlZFSpPECtOAPFXDA1VxUY7Jv41oYBqxHUHQr9_K0WN3E11P0Zbsr-r3nK6PIFEL5pWeve0xLaBdCjpXMfkcVsAz8gsFXy4ZDyhTuUFd-CTOd2Bt_wAZZVXVjjCSUiXj1ncGearq_V342g0xy9xRpsl1MwyTjn8nmrMmFC5DAkYo8vqZFQpY5i9mVnEKZCRA/Annotation%202019-06-11%20113615.png)


**How is this information encoded in the ETCBC database?**
![](https://puz9rq.dm.files.1drv.com/y4maSHxQ9-k0iwUVUa1_RmIdxilEmHk7wQAOI8S4pBWit7qcDEeoW11GYwN13JEd1WLuRqq5U6b27_C9ClH9NCYEeAgF1SvDf5bKbDCm35OcgxW6SalHdTWtcgZQlOaJMCfv4Twb422x603fMJuoz9SAHUvSrT4DnhPBNJ4vcMucv2MUjVltX24iCJ1zzEXj4vp-yVwWG0eB28GRykUErQbRA/Annotation%202019-06-11%20114149.png)

### Where are the Codes?
Now we have understood the basics of how the database is built we now need to figure out what codes/abbreviations are available for `object types` (e.g. word), for `features` (e.g. number), and for values (e.g. s, p).

A full list of features codes can be found here: https://etcbc.github.io/bhsa/features/0_home/

Check the table of contents on the right to get quick access to all linguistc levels with a desription of their `features` and `values`:
![](https://erkboa.dm.files.1drv.com/y4mK9vhAUcxzi8kXo4iYtH-sj1AFhc4j_3gfUgu9yz6alI4ak4NTrHgek5Y0r_xamniEXKUubzQLjUx62fdbORlf5eWdIJRhUKuXw9-SViXSfcyufvx0NLKyBuH8dNNRiDDJUcracvikdoTjQBEVN5VMHospwbXv3FfH6ztAfDEJu0WC0rr7AzwosO0WQt4opHyeOpRH0Zby3nZnqvoaS5xFQ/Annotation%202019-06-11%20123203.png)

Often it is more time efficient if you just look up your codes via SHEBANQ since all features and values are hyperlinked. You can get to the codes in SHEBANQ by clicking on the verse number and select from the **legend** the codes you need for your query:
![](https://d0r0lq.dm.files.1drv.com/y4mkfdft1bmjTrPqv1UwQV2fviFGALWL5_ZUxgyjvF5s0pPJkmE8-hejcn_7GIa2LBllStPItKwLMFBtfOyHQ0WEsY7RVByW6W-QDxSBIX3ttNEwEjO6azZbXvEHS0Z2C2ZjuUk2UVS-ng9d6RpqmQnVzfkIeyjjkcca6FW9sVoLtlBwJv_U8f64d5kJe__hx3FutOE4PPDi-fQPyMn4a197A/Annotation%202019-06-11%20124142.png)


In [39]:
valence1='''
verse book=Genesis chapter=20
    phrase function=Objc
'''
valence1  = A.search(valence1)
A.table(valence1, start=1, end=2, condensed=True)

  0.35s 19 results


# Searching for Verbal Valence
Valence searches help you to find the meaning of words, particularly words, by studying the relatations between verbs to their complements (necessary elements). Complements are phrase that are necessary for verbs to activate a certain meaning (think of "shop" activates in "Peter runs a shop" a different meaning for "run" than "marathon" does in "Peter runs a marathon".
Lets do some valence searches for the verb LQX[ in Genesis 20:2.

Lets first have a look at the linguistc structure of Gen 20:2 in SHEBANQ:
![](https://3mqm8g.dm.files.1drv.com/y4pN6YrP0nK8ObHEmeBX_BtnWqYx-2eR6kIbz_ezE7yd_iSYXdxIm_k4PCSYDNWCFiGO57mj8ubgByErOgRDWHgZx5B1I47Wm2vFs8Hf1OdoWh9OnUh4g_QgPL_beB3Pcpzlz1ecRnr4UmQt-fuWthdgeOn5WMFGuDfhfZ2beiUdbO8A1bgitFZNw-xS3rjmU3-cdwpsj9dTA4Z2YB8WpmZKQ/Annotation%202019-06-11%20125345.png)

From this structure we can easily built our first valence sensitive query:

In [32]:
valence1='''
clause
    phrase function=Pred
        word lex=LQX[
    phrase function=Objc
'''
valence1  = A.search(valence1)
A.table(valence1, start=1, end=2, condensed=True)

  1.75s 708 results


Obviously, LQX[ takes often a direct object (708 times).

But how often is that direct object a female proper name (think of Sara)?

In [33]:
valence2='''
clause
    phrase function=Pred
        word lex=LQX[
    phrase function=Objc
        word sp=nmpr gn=f nu=sg
'''
valence2  = A.search(valence2)
A.table(valence2, start=1, end=24, condensed=True)

  2.20s 24 results


Interstingly, where the female proper name is the only element int he object phrase (e.g. Gen 28:9) a complement phrase with L+>CH follows usually (e.g. Gen 28:9, 34:4, etc.).
Lets search for those cases so that we can get a better picture of what is going on.

In [34]:
# task 4
# Alter the above query in such a way that you are also looking for complement phrases that contain the noun >CH/.
# What is the meaning of that construction? How would you best render this particular LQX valence in English?

In [37]:
test1='''
clause
    phrase function=Pred
        word lex=LQX
    phrase function=Cmpl
        word lex=>CH/
        
'''
test1  = A.search(test1)
A.table(test1, start=1, end=24, condensed=True)

  1.01s 144 results
